pkg/cluster/updater.go

```diff
@@ -0,0 +1,372 @@
+// in the format of $repo:$tag.
+func (ci *ContainerImage) String() string {
+	return fmt.Sprintf("%s:%s", ci.Repo, ci.Tag)
+}
+
+// UpdaterFunc is a function that can be used to update a
+// single component of a certain type. A concrete updater
+// function may be something like 'DaemonSetUpdaterFunc`.
+type ComponentUpdaterFunc func(clientset.Interface, string, *ContainerImage) error
+
+// NewClusterUpdater returns a ClusterUpdater struct with defaults.
+func NewClusterUpdater(client clientset.Interface, newVersion, oldVersion *ContainerImage, config *v1.ConfigMap) (*ClusterUpdater, error) {
+	b := record.NewBroadcaster()
+	r := b.NewRecorder(api.EventSource{
+		Component: "cluster-updater",
+	})
+	p, err := client.Core().Pods(api.NamespaceSystem).Get(os.Getenv("MY_POD_NAME"))
+	if err != nil {
+		return nil, fmt.Errorf("error getting self pod: %v", err)
+	}
+	return &ClusterUpdater{
+		Client:     client,
+		Components: DefaultComponentUpdaterList,
+		NewVersion: newVersion,
+		OldVersion: oldVersion,
+		RollbackFn: DefaultRollback,
+		Events:     r,
+		SelfObj:    p,
+		Config:     config,
+	}, nil
+}
+
+// DaemonsetRollingUpdate will perform a safe rolling update on the DaemonSet
+// specified by `name`. An error will be returned if the rolling update failed.
+func DaemonsetRollingUpdate(client clientset.Interface, name string, image *ContainerImage) error {
+	ds, err := client.Extensions().DaemonSets(api.NamespaceSystem).Get(name)
+	if err != nil {
+		return err
+	}
+	// Create new DS.
+	ds.Labels["version"] = image.Tag
+	ds.Spec.Template.Labels["version"] = image.Tag
+	for i, c := range ds.Spec.Template.Spec.Containers {
+		if c.Name == name {
+			glog.Infof("updating image for container: %s", c.Name)
+			ds.Spec.Template.Spec.Containers[i].Image = image.String()
+			break
+		}
+	}
+	ds, err = client.Extensions().DaemonSets(api.NamespaceSystem).Update(ds)
+	if err != nil {
+		return err
+	}
+	l, err := labels.Parse(fmt.Sprintf("k8s-app=%s", name))
+	if err != nil {
+		return err
+	}
+	lo := api.ListOptions{
+		LabelSelector: l,
+	}
+	pl, err := client.Core().Pods(api.NamespaceSystem).List(lo)
+	if err != nil {
+		return err
+	}
+	for _, p := range pl.Items {
+		// Delete old DS Pod.
+		glog.Infof("Deleting pod %s", p.Name)
+		err = client.Core().Pods(api.NamespaceSystem).Delete(p.Name, nil)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		glog.Infof("Deleted pod %s", p.Name)
+		// Wait for new DS Pod.
+		for {
+			glog.Info("Waiting for new pod to come up...")
+			upl, err := client.Core().Pods(api.NamespaceSystem).List(lo)
+			if err != nil {
+				// Do not return an error here, since we have just deleted a pod,
+				// it is quite possible we have deleted an API Server pod. We may fail
+				// to contact it for a bit until the checkpointer API Server takes over.
+				glog.Error(err)
+			} else {
+				if ds.Status.DesiredNumberScheduled == int32(len(upl.Items)) {
+					if allContainersRunning(upl) {
```

Comment:
You could use upstream logic here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L629

Main differences are looking at the pod status transition time (I'm assuming in case the last status was never updated) and looking at podReady condition: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L651